utils/src/mem_size.rs

```diff
@@ -0,0 +1,117 @@
+//! Defines a [`MemSizeEstimator`] trait and a companying [`MemMode`] which are used to
+//! estimate sizes of run-time objects in memory including deep heap allocations. See
+//! struct-level docs for moew details.
+
+use std::{collections::HashSet, mem::size_of, sync::Arc};
+
+use parking_lot::RwLock;
+
+/// The memory mode of the tracked object
+#[derive(Debug, Clone, Copy)]
+pub enum MemMode {
+    Bytes,
+    Units,
+}
+
+/// The contract for estimating deep memory size owned by this object. Implementors
+/// are expected to support only a single function - bytes or units. Objects with pre-compliation
+/// known static size or which are containers of items with static size should implement the `_units`
+/// estimation and return the number of logical items (usually 1 or the number of items in
+/// the container). Objects with varying runtime sizes should implement the `_bytes` estimation.
+///
+/// By panicking on the remaining unimplemented function we ensure that tests will catch any inconsistency over the
```

Comment:
> By panicking on the remaining unimplemented function

The remaining function is not necessarily always unimplemented (such as in the implementations for `Arc<T>` and `RwLock<T>` below). Please add a comment about in what instances it would be acceptable that both are implemented.

Reply:
Arc and RwLock are simple delegations. The underlying object is expected to implement only a single method. I'll comment